### PR TITLE
Fix issue #329

### PR DESCRIPTION
**Pull request by reicheratwork**
Issue #329 was caused by default constructed and unset QoS Policies being indistinguishable for CDDS-CXX, which caused issues when CDDS-C would fill in certain fields based on other information if the Policy wasn't set. Such DataRepresentation in issue #329.
The solution was to make the C++ QoS implementations mirror the C sparse QoSes.
Policies that are unset in a C++ QoS are now NOT exported to the C QoS representation, wherease they previously always were.
Also, the contents of default constructors reader/writer/topic/publisher/subscriber QoS are now taken from CDDS-C, in stead of being explicitly defined, removing a double definition. C++'s default QoSes will now mirror C's default QoSes.

@javiersorianoruiz If would be so kind as to help with the review process?

### CODE REVIEW

File: src/ddscxx/src/org/eclipse/cyclonedds/pub/qos/DataWriterQosDelegate.cpp [hunk 1]
```diff
@@ -33,98 +35,111 @@ namespace qos
 
 DataWriterQosDelegate::DataWriterQosDelegate()
 {
-    reliability_.kind(dds::core::policy::ReliabilityKind::RELIABLE);
+    ddsc_qos(&ddsi_default_qos_writer);
+    present() &= ~QP_DATA_REPRESENTATION;
+    check();
+    policy(dds::core::policy::Reliability(dds::core::policy::ReliabilityKind::RELIABLE));
```
Comment (eboasson) [inline src/ddscxx/src/org/eclipse/cyclonedds/pub/qos/DataWriterQosDelegate.cpp:41]: Isn't this covered by `ddsc_qos(&ddsi_default_qos_writer)`?

File: src/ddscxx/src/org/eclipse/cyclonedds/sub/qos/DataReaderQosDelegate.cpp [hunk 2]
```diff
@@ -249,18 +301,29 @@ DataReaderQosDelegate::operator==(const DataReaderQosDelegate& other) const
...
-    durability_  = tqos.policy<dds::core::policy::Durability>();
-    deadline_    = tqos.policy<dds::core::policy::Deadline>();
-    budget_      = tqos.policy<dds::core::policy::LatencyBudget>();
-    liveliness_  = tqos.policy<dds::core::policy::Liveliness>();
-    reliability_ = tqos.policy<dds::core::policy::Reliability>();
-    order_       = tqos.policy<dds::core::policy::DestinationOrder>();
-    history_     = tqos.policy<dds::core::policy::History>();
-    resources_   = tqos.policy<dds::core::policy::ResourceLimits>();
-    ownership_   = tqos.policy<dds::core::policy::Ownership>();
+    if (tqos.present() & QP_DURABILITY)
```
Comment (eboasson) [inline src/ddscxx/src/org/eclipse/cyclonedds/sub/qos/DataReaderQosDelegate.cpp:304]: This should be looking at the `DURABILITY` flag for all settings

File: src/ddscxx/include/org/eclipse/cyclonedds/domain/qos/DomainParticipantQosDelegate.hpp [hunk 3]
```diff
@@ -20,6 +20,8 @@
 
 #include <dds/core/policy/CorePolicy.hpp>
 
+#include <dds/ddsi/ddsi_xqos.h>
```
Comment (eboasson) [inline src/ddscxx/include/org/eclipse/cyclonedds/domain/qos/DomainParticipantQosDelegate.hpp:23]: I suspect this header file gets included by the application code as well, and if so, then including this header file pollutes the name space rather badly. It isn't really part of the public API, but language bindings are a bit special, and this does save quite a few lines of boilerplate compared to calling `dds_qget_` functions all the time and checking the return value.

The only thing you actually use from it (unless I overlooked something) are the `QP_` flags, so a the name space pollution could be addressed by reorganising the C header files a bit, but I think I'd be happier if the C++ binding used its own set of (namespaced) constants, with a single `.cpp` file including the C header file using static assertions to verify that the bits are actually the same.

Best would of course be if there's a way of decoupling it completely without adding a lot of boilerplate.

File: src/ddscxx/include/org/eclipse/cyclonedds/domain/qos/DomainParticipantQosDelegate.hpp [hunk 4]
```diff
@@ -36,11 +38,6 @@ namespace qos
 class OMG_DDS_API DomainParticipantQosDelegate
 {
 public:
-    DomainParticipantQosDelegate();
```
Comment (eboasson) [inline src/ddscxx/include/org/eclipse/cyclonedds/domain/qos/DomainParticipantQosDelegate.hpp:39]: I surprises me you can just throw these away, are you sure that can't cause any problems?

File: src/ddscxx/include/org/eclipse/cyclonedds/domain/qos/DomainParticipantQosDelegate.hpp [hunk 5]
```diff
@@ -56,11 +53,13 @@ class OMG_DDS_API DomainParticipantQosDelegate
...
-    DomainParticipantQosDelegate& operator =(const DomainParticipantQosDelegate& other);
 
+    const uint64_t &present() const {return present_;}
+    uint64_t &present() {return present_;}
 private:
-    dds::core::policy::UserData user_data_;
-    dds::core::policy::EntityFactory entity_factory_;
+    uint64_t                          present_ = 0;
+    dds::core::policy::UserData       user_data_;
+    dds::core::policy::EntityFactory  entity_factory_;
```
Comment (eboasson) [inline src/ddscxx/include/org/eclipse/cyclonedds/domain/qos/DomainParticipantQosDelegate.hpp:62]: I tried to find out whether the default constructor for the various policies will initialise the value properly, but I am not quite sure that it really does: for example, `EntityFactory` seems to end up in `PolicyDelegate.hpp:222` with the default-default constructor, and IINM, that means the boolean in there won't be initialized. (Or perhaps that was for POD and classes are different?)

I don't think there have been any changes to this aspect, but I think it is worth double-checking.

File: src/ddscxx/include/org/eclipse/cyclonedds/domain/qos/DomainParticipantQosDelegate.hpp [hunk 6]
```diff
@@ -92,6 +92,7 @@ template<>
 inline dds::core::policy::EntityFactory&
 DomainParticipantQosDelegate::policy<dds::core::policy::EntityFactory> ()
 {
+    present_ |= QP_ADLINK_ENTITY_FACTORY;
```
Comment (eboasson) [inline src/ddscxx/include/org/eclipse/cyclonedds/domain/qos/DomainParticipantQosDelegate.hpp:95]: This I don't really understand: this means that reading the value marks it as being set, which in turn means it gets passed on to the C API, right? That may or may not be wise (I can see arguments either way), but having two getters and only one of them setting the flag seems like a recipe for confusion. Given that the `const` variant can't set the flag, I think that means the non-const version also mustn't set the flag.

Comment (reicheratwork) [reply src/ddscxx/include/org/eclipse/cyclonedds/domain/qos/DomainParticipantQosDelegate.hpp:95]: The idea here was that we are passing a reference to the EntityFactory QoS, so the person receiving the reference can modifies members of the EntityFactory QoS and hereby implicitly set the presence flag for the EntityFactory QoS.

File: src/ddscxx/include/org/eclipse/cyclonedds/domain/qos/DomainParticipantQosDelegate.hpp (hunk 4 above)
Comment (reicheratwork) [reply src/ddscxx/include/org/eclipse/cyclonedds/domain/qos/DomainParticipantQosDelegate.hpp:39]: Well, the copy constructor and destructor are in my eyes the definition of boilerplate, and can be completely handed off to the compiler.
By setting the default values of the members, which do not have default constructors (like the present_ flag on line 60), these values will then be passed on during construction when no other values are supplied.
This allows the user to skip the definition of the default constructor and just fall back on the default values supplied in the definition of the class.

File: src/ddscxx/include/org/eclipse/cyclonedds/domain/qos/DomainParticipantQosDelegate.hpp (hunk 5 above)
Comment (reicheratwork) [reply src/ddscxx/include/org/eclipse/cyclonedds/domain/qos/DomainParticipantQosDelegate.hpp:62]: I will double check this.

File: src/ddscxx/src/org/eclipse/cyclonedds/pub/qos/DataWriterQosDelegate.cpp (hunk 1 above)
Comment (reicheratwork) [reply src/ddscxx/src/org/eclipse/cyclonedds/pub/qos/DataWriterQosDelegate.cpp:41]: Let me double check.

File: src/ddscxx/src/org/eclipse/cyclonedds/sub/qos/DataReaderQosDelegate.cpp (hunk 2 above)
Comment (reicheratwork) [reply src/ddscxx/src/org/eclipse/cyclonedds/sub/qos/DataReaderQosDelegate.cpp:304]: OUCH!!!
I will fix this immediately!!!

File: src/ddscxx/src/org/eclipse/cyclonedds/pub/qos/DataWriterQosDelegate.cpp (hunk 1 above)
Comment (reicheratwork) [reply src/ddscxx/src/org/eclipse/cyclonedds/pub/qos/DataWriterQosDelegate.cpp:41]: As we just use ddsc_qos, and not ddsi_xqos_mergein_missing, we do need to "unset" the DataRepresentation policy.
Setting of the Reliability policy is not necessary, as that set through ddsc_qos, i will remove this.

File: src/ddscxx/include/org/eclipse/cyclonedds/domain/qos/DomainParticipantQosDelegate.hpp (hunk 5 above)
Comment (reicheratwork) [reply src/ddscxx/include/org/eclipse/cyclonedds/domain/qos/DomainParticipantQosDelegate.hpp:62]: The old iteration with the explicit default constructor also takes the default constructor for the EntityFactory QoSPolicy.
Maybe fix it at the default constructor for the QoSPolicy?

**Comment by reicheratwork** [comment]
Updated the PR, but now it cannot be merged until [CycloneDDS-C PR 1503](https://github.com/eclipse-cyclonedds/cyclonedds/pull/1503) is merged, as it exposes the ``ddsi_default_qos_participant`` which this PR requires.

**Review by eboasson — APPROVED** [review]
LGTM! And thank you for updating it several times while I kept merging stuff into the core library breaking this several times ☺️

**Comment by javiersorianoruiz** [comment]
I have downloaded and installed current master versions of cyclonedds, cyclonedds-python and cyclonedds-cxx and my XTypes test for Cpp and Python are working now properly.

More info about these tests:

https://github.com/javiersorianoruiz/cyclonedds-xtypes-testing

**Pull request merged**